File: src/RideSharing.Infrastructure/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using RideSharing.Application.Common;
using RideSharing.Application.DTOs.Users;
using RideSharing.Application.Interfaces;
using RideSharing.Domain.Enums;
using RideSharing.Persistence.Context;

namespace RideSharing.Infrastructure.Services;

public class UserService : IUserService
{
    private readonly RideSharingDbContext _db;

    public UserService(RideSharingDbContext db) => _db = db;

    public async Task<ApiResponse<UserProfileDto>> GetMyProfileAsync(Guid userId)
    {
        var user = await _db.Users
            .Include(u => u.Institution)
            .Include(u => u.UserRoles).ThenInclude(ur => ur.Role)
            .FirstOrDefaultAsync(u => u.Id == userId && !u.IsDeleted);

        if (user == null)
            return ApiResponse<UserProfileDto>.FailureResponse("User not found.");

        return ApiResponse<UserProfileDto>.SuccessResponse(Map(user));
    }

    public async Task<ApiResponse<UserProfileDto>> UpdateMyProfileAsync(Guid userId, UpdateProfileRequest request)
    {
        var user = await _db.Users
            .Include(u => u.Institution)
            .Include(u => u.UserRoles).ThenInclude(ur => ur.Role)
            .FirstOrDefaultAsync(u => u.Id == userId && !u.IsDeleted);

        if (user == null)
            return ApiResponse<UserProfileDto>.FailureResponse("User not found.");

        if (request.InstitutionId.HasValue)
        {
            var exists = await _db.Institutions.AnyAsync(i => i.Id == request.InstitutionId.Value && !i.IsDeleted);
            if (!exists)
                return ApiResponse<UserProfileDto>.FailureResponse("Institution not found.");
        }

        user.FirstName = request.FirstName.Trim();
        user.LastName = request.LastName.Trim();
        user.PhoneNumber = request.PhoneNumber;
        user.Gender = (Gender)request.Gender;
        user.DateOfBirth = request.DateOfBirth;
        user.InstitutionId = request.InstitutionId;
        user.UpdatedAt = DateTime.UtcNow;

        await _db.SaveChangesAsync();

        // reload institution name
        if (user.InstitutionId.HasValue)
            await _db.Entry(user).Reference(u => u.Institution).LoadAsync();

        return ApiResponse<UserProfileDto>.SuccessResponse(Map(user), "Profile updated.");
    }

    public async Task<ApiResponse<UserProfileDto>> UpdateProfileImageAsync(Guid userId, string imageUrl)
    {
        var user = await _db.Users
            .Include(u => u.Institution)
            .Include(u => u.UserRoles).ThenInclude(ur => ur.Role)
            .FirstOrDefaultAsync(u => u.Id == userId && !u.IsDeleted);

        if (user == null)
            return ApiResponse<UserProfileDto>.FailureResponse("User not found.");

        user.ProfileImageUrl = imageUrl;
        user.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();

        return ApiResponse<UserProfileDto>.SuccessResponse(Map(user), "Profile image updated.");
    }

    private static UserProfileDto Map(Domain.Entities.User user) => new()
    {
        Id = user.Id,
        FirstName = user.FirstName,
        LastName = user.LastName,
        Email = user.Email,
        PhoneNumber = user.PhoneNumber,
        Gender = user.Gender.ToString(),
        DateOfBirth = user.DateOfBirth,
        ProfileImageUrl = user.ProfileImageUrl,
        IsVerified = user.IsVerified,
        IsActive = user.IsActive,
        InstitutionId = user.InstitutionId,
        InstitutionName = user.Institution?.Name,
        Roles = user.UserRoles?.Select(ur => ur.Role.Name).ToList() ?? new()
    };
}

File: RideSharing.Web/src/app/core/services/user.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../models/auth.models';

export interface UserProfile {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber?: string;
  gender: string;
  dateOfBirth?: string;
  profileImageUrl?: string;
  isVerified: boolean;
  isActive: boolean;
  institutionId?: string;
  institutionName?: string;
  roles: string[];
}

export interface UpdateProfileRequest {
  firstName: string;
  lastName: string;
  phoneNumber?: string;
  gender: number;
  dateOfBirth?: string | null;
  institutionId?: string | null;
}

@Injectable({ providedIn: 'root' })
export class UserService {
  private http = inject(HttpClient);
  private base = `${environment.apiUrl}/users`;

  getMe(): Observable<ApiResponse<UserProfile>> {
    return this.http.get<ApiResponse<UserProfile>>(`${this.base}/me`);
  }

  updateMe(body: UpdateProfileRequest): Observable<ApiResponse<UserProfile>> {
    return this.http.put<ApiResponse<UserProfile>>(`${this.base}/me`, body);
  }

  uploadProfileImage(file: File): Observable<ApiResponse<UserProfile>> {
    const form = new FormData();
    form.append('file', file);
    return this.http.post<ApiResponse<UserProfile>>(`${this.base}/profile-image`, form);
  }
}
